src/mon/FSCommands.cc

```diff
@@ -1245,6 +1242,125 @@ class RenameFilesystemHandler : public FileSystemCommandHandler
   Paxos *m_paxos;
 };
 
+class SwapFilesystemHandler : public FileSystemCommandHandler
+{
+  public:
+  explicit SwapFilesystemHandler(Paxos *paxos)
+    : FileSystemCommandHandler("fs swap"), m_paxos(paxos)
+  {
+  }
+
+  bool batched_propose() override {
+    return true;
+  }
+
+  int handle(
+      Monitor *mon,
+      FSMap& fsmap,
+      MonOpRequestRef op,
+      const cmdmap_t& cmdmap,
+      std::ostream &ss) override
+  {
+    ceph_assert(m_paxos->is_plugged());
+
+    // Check for confirmation flag
+    bool confirmation_flag = false;
+    cmd_getval(cmdmap, "yes_i_really_mean_it", confirmation_flag);
+    if (!confirmation_flag) {
+      ss << "This is a potentially disruptive operation, client\'s cephx "
+	"credentials may need to be reauthorized to access the file systems "
+	"and its pools. Add --yes-i-really-mean-it if you are sure you wish "
+	"to continue.";
+      return -EPERM;
+    }
+
+    string fs1_name, fs2_name;
+    bool swap_flag;
+    cmd_getval(cmdmap, "fs1_name", fs1_name);
+    cmd_getval(cmdmap, "fs2_name", fs2_name);
+    cmd_getval(cmdmap, "swap_fscids", swap_flag);
+    auto fs1 = fsmap.get_filesystem(fs1_name);
+    auto fs2 = fsmap.get_filesystem(fs2_name);
+
+    // Check that CephFSs for both gives names exists
+    if (fs1 == nullptr or fs2 == nullptr) {
+      if (fs1 == nullptr and fs2 != nullptr) {
+	ss << "File system \"" << fs1_name << "\" doesn\'t exist on this "
+	      "Ceph cluster.";
+	return -ENOENT;
+      } else if (fs1 != nullptr and fs2 == nullptr) {
+	ss << "File system \"" << fs2_name << "\" doesn\'t exist on this "
+	      "Ceph cluster.";
+	return -ENOENT;
+      } else {
+	ss << "Neither file system \"" << fs1_name << "\" nor file "
+	      "system \"" << fs2_name << "\" exists on this Ceph cluster.";
+	return -ENOENT;
+      }
+    }
+
+    // Check that CephFS mirroring for both CephFSs is disabled.
+    if (fs1->mirror_info.mirrored or fs2->mirror_info.mirrored) {
+      if (fs1->mirror_info.mirrored) {
+	ss << "Mirroring is enabled on file system \""<< fs1_name <<
+	  "\". Disable mirroring on the file system after ensuring it's OK to "
+	  "do so, and then retry to swap.";
+	return -EPERM;
+      } else if (fs2->mirror_info.mirrored) {
+	ss << "Mirroring is enabled on file system \""<< fs2_name <<
+	"\". Disable mirroring on the file system after ensuring it's OK to "
+	"do so, and then retry to swap.";
+	return -EPERM;
+      } else {
+	ss << "Mirroring is enabled on file system1 \""<< fs1_name <<
+	      "\" as well as on \"" << fs2_name << "\". Disable mirroring on "
+	      "both the file systems after ensuring it's OK to do so, and then "
+	      "retry to swap.";
+	return -EPERM;
+      }
+    }
+
+    if (!mon->osdmon()->is_writeable()) {
+      // not allowed to write yet, so retry when we can
+      mon->osdmon()->wait_for_writeable(op, new PaxosService::C_RetryMessage(mon->mdsmon(), op));
+      return -EAGAIN;
+    }
```

Comment:
shouldn't this check be done before the code to check for FSs existence and mirroring?

Reply:
There's no point in waiting when FSs doesn't exist or mirroring is enabled.